### ISSUE TITLE
Firebase Tools 6.10.0 doesn't serve OPTIONS requests and overwrites CORS

### ISSUE THREAD

**Issue by jaschaio** [firebase/firebase-tools#1342]
### Environment info

Platform: macOSX 10.14.5
Node: 10.15.3
firebase-tools: 6.10.0
firebase-functions: 2.3.1
firebase-admin: 8.0.0

### Test case

A freshly generated firebase project using `firebase init` with functions.

Use Node `10.15.3` and change the `package.json` engines node to `10`.

Create a `.nvmrc` file that contains `v10.15.3`

Install `firebase-admin@8.0.0` and `firebase-functions@2.3.1` within the functions directory. 

Make sure firebase-tools is running `6.10.0` by running `firebase --version`.

Run `npm install --save grpc` within the functions direcotry to prevent #1341

Than install express and cors `npm i express cors` as well within the functions directory.

Replace the `index.js` with the following content:

```javascript
const functions = require( 'firebase-functions' );
const express = require( 'express' );
const cors = require( 'cors' );

// Create express app
const app = express();

// Use CORS middleware
app.use( cors( {
    origin: 'https://example.org',
    allowedHeaders: 'X-Something-Special,Authorization',
    credentials: true,
    optionsSuccessStatus: 204
} ) );

// Setup dummy post Route
app.post( '/', ( req, res ) => {

    res.send( 200 );

} );

// Export express app via helloWorld method
exports.helloWorld = functions.https.onRequest( app );
```

### Steps to reproduce

Run `firebase serve` within the newly created project as described above and make an options request to the exported http function using Postman or CURL:

```
curl -H "Origin: http://example.com" \
  -H "Access-Control-Request-Method: POST" \
  -H "Access-Control-Request-Headers: X-Requested-With" \
  -X OPTIONS --verbose \
  http://localhost:5000/example-app/us-central1/helloWorld
```
(replace your project name for the right url) 

The response CORS Headers are apparently being set by firebase-tools instead of the CORS middleware. Its like options request get completely ignored and handled by firebase-tools instead of the actual express app.

**Downgrading to `firebase-tools@6.8.0` resolves the issue.**

### Expected behavior

Setting the correct CORS Headers as defined by the cors module:

```
x-powered-by →Express
cache-control →no-cache, no-store, must-revalidate
pragma →no-cache
expires →0
access-control-allow-origin →https://example.org
vary →Origin
access-control-allow-credentials →true
access-control-allow-methods →GET,HEAD,PUT,PATCH,POST,DELETE
access-control-allow-headers →X-Something-Special,Authorization
content-length →0
date →Wed, 29 May 2019 11:09:49 GMT
connection →close
```

By the way this has nothing to do with the cors module. Even when defining an options route via `app.options( '/', /* ... */ )` and setting the headers manually they are overwritten.

### Actual behavior

Instead completely arbitrary CORS Headers are being set:

```
X-Powered-By →Express
Access-Control-Allow-Origin →*
Access-Control-Allow-Headers →Origin, X-Requested-With, Content-Type, Authorization, Accept
Access-Control-Allow-Methods →GET,OPTIONS,POST
Allow →POST,GET,HEAD
Content-Type →text/html; charset=utf-8
Content-Length →13
ETag →W/"d-BVyHhJ9/eETU1VqxhQuV1zTNV1Q"
Date →Wed, 29 May 2019 11:14:05 GMT
Connection →keep-alive
```

**Comment by jaschaio** [comment]
#1325 seems related, but the fix doesn't seem to apply to `OPTIONS` requests.

I tried exporting this method as described in the fix https://github.com/firebase/firebase-tools/pull/1332:

```
exports.httpFn = functions.https.onRequest((request, response) => {
    response.send({
        method: request.method,
        path: request.path, 
        query: request.query 
    });
});
```

And a `OPTIONS` request doesn't even triggers the function, so it never arrives. So this has nothing to do with express or cors, but how firebase-tools handles options requests.

**Comment by bjon** [comment]
I'm experiencing the same problem

**Comment by mbleigh** [comment]
@samtstern @abeisgoat can you take a look or triage this?

**Comment by samtstern** [comment]
The change in https://github.com/firebase/firebase-tools/pull/1335 will fix this and will be included in the next bugfix release, thanks for your patience!

**Comment by QuantumInformation** [comment]
Not sure if this is the same issue but if I try Access this with the emulator
```js
exports.listUsers = functions.https.onRequest(async (req, res) => {
    res.set("Access-Control-Allow-Origin", "*");
    const data = await users_1.getSomeUsers(10);
    res.json(data);
});
```

I end up with

Access to fetch at 'http://localhost:5002/svelte-fullstack-starter/us-central1/listUsers' from origin 'http://localhost:5001' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.

**Comment by samtstern** [comment]
@QuantumInformation this issue was filed about version `6.10.0` but the latest is two major versions after that (`8.1x.x`).  If you can reproduce this on the latest release please file a new issue,

**Comment by QuantumInformation** [comment]
ok, sorry if I dug up a can of worms
